wallet/src/store.rs

```diff
@@ -0,0 +1,1196 @@
+                    batch.delete_cell_out_point_lock_hash(&lock_hash_index.cell_out_point);
+                });
+
+            let iter = self
+                .db
+                .iter(COLUMN_LOCK_HASH_TRANSACTION, lock_hash.as_bytes())
+                .expect("wallet db iter should be ok");
+
+            iter.take_while(|(key, _)| key.starts_with(lock_hash.as_bytes()))
+                .for_each(|(key, _)| {
+                    let lock_hash_index = LockHashIndex::from_slice(&key);
+                    batch.delete_lock_hash_transaction(&lock_hash_index);
+                });
+
+            batch.delete_lock_hash_index_state(&lock_hash);
+        });
+    }
+}
+
+impl<CS: ChainStore + 'static> DefaultWalletStore<CS> {
+    pub fn new(db_config: &DBConfig, shared: Shared<CS>) -> Self {
+        let db = RocksDB::open(db_config, COLUMNS);
+        DefaultWalletStore {
+            db: Arc::new(db),
+            shared,
+        }
+    }
+
+    pub fn start<S: ToString>(self, thread_name: Option<S>, notify: &NotifyController) {
+        let mut thread_builder = thread::Builder::new();
+        if let Some(name) = thread_name {
+            thread_builder = thread_builder.name(name.to_string());
+        }
+
+        let new_tip_receiver = notify.subscribe_new_tip(WALLET_STORE_SUBSCRIBER);
+        thread_builder
+            .spawn(move || loop {
+                select! {
+                    recv(new_tip_receiver) -> msg => match msg {
+                        Ok(tip_changes) => self.update(&tip_changes.detached_blocks, &tip_changes.attached_blocks),
+                        _ => {
+                            error!(target: "wallet", "new_tip_receiver closed");
+                            break;
+                        }
+                    },
+                }
+            })
+            .expect("start DefaultWalletStore failed");
+    }
+
+    // helper function
+    fn commit_batch<F>(&self, process: F)
+    where
+        F: FnOnce(&mut WalletStoreBatch),
+    {
+        match self.db.batch() {
+            Ok(batch) => {
+                let mut batch = WalletStoreBatch {
+                    batch,
+                    insert_buffer: HashMap::new(),
+                    delete_buffer: HashSet::new(),
+                };
+                process(&mut batch);
+                batch.commit();
+            }
+            Err(err) => {
+                error!(target: "wallet", "wallet db failed to create new batch, error: {:?}", err);
+            }
+        }
+    }
+
+    pub fn sync_index_states(&self) {
+        info!(target: "wallet", "Start sync index states with chain store");
+        let mut lock_hash_index_states = self.get_lock_hash_index_states();
+        if lock_hash_index_states.is_empty() {
+            return;
+        }
+        // need to lock chain state, avoids inconsistent state in processing
+        let chain_state = self.shared.lock_chain_state();
```

Comment:
Like the previous one, can we periodically release and re-acquire the lock to give other threads requesting the lock a chance to succeed?